spring-ai-core/src/main/java/org/springframework/ai/chat/prompt/Prompt.java

```diff
@@ -92,4 +96,31 @@ public int hashCode() {
 		return Objects.hash(this.messages, this.modelOptions);
 	}
 
+	public Prompt copy() {
+		return new Prompt(instructionsCopy(), this.modelOptions);
+	}
+
+	private List<Message> instructionsCopy() {
+		List<Message> messagesCopy = new ArrayList<>();
+		this.messages.forEach(message -> {
+			if (message instanceof UserMessage) {
+				messagesCopy.add(new UserMessage(message.getContent(), message.getMedia(), message.getMetadata()));
+			}
+			else if (message instanceof SystemMessage) {
+				messagesCopy.add(new SystemMessage(message.getContent()));
+			}
+			else if (message instanceof AssistantMessage) {
+				messagesCopy.add(new AssistantMessage(message.getContent(), message.getMetadata()));
+			}
+			else if (message instanceof FunctionMessage) {
+				messagesCopy.add(new FunctionMessage(message.getContent(), message.getMetadata()));
+			}
+			else {
+				throw new IllegalArgumentException("Unsupported message type: " + message.getClass().getName());
+			}
```

Comment:
There are several areas in the code where the list of messages are being filtered. Would it make sense to introduce a map representation of messages by type? Perhaps something like: 
```
Map<Class<? extends Message>, List<Message>> messagesByType
```

Reply:
Not sure I completely undestand the suggestion?  Where is this map going to be? Also the order of the messages is important and must be preserved (e.g. User followed by Assistant then another pair of user/assistant ... ).

Reply:
Using a `LinkedHashMap` as the Map implementation would preserve ordering. I was suggesting this in place of or in addition to the `List<Message>` as part of `Prompt`.